File: app/csharp/AccountingSystem.Api/Domain/Audit/AuditLog.cs
namespace AccountingSystem.Domain.Audit;

/// <summary>
/// 監査ログドメインモデル（不変）
/// </summary>
public record AuditLog
{
    public long? Id { get; init; }
    public required string EntityType { get; init; }
    public required string EntityId { get; init; }
    public required AuditAction Action { get; init; }
    public required string UserId { get; init; }
    public required string UserName { get; init; }
    public required DateTime Timestamp { get; init; }
    public Dictionary<string, object>? OldValues { get; init; }
    public Dictionary<string, object>? NewValues { get; init; }
    public Dictionary<string, object>? Changes { get; init; }
    public string? Reason { get; init; }
    public string? IpAddress { get; init; }
    public string? UserAgent { get; init; }

    /// <summary>
    /// ファクトリメソッド：CREATE 操作用
    /// </summary>
    public static AuditLog Create(
        string entityType,
        string entityId,
        AuditAction action,
        string userId,
        string userName,
        Dictionary<string, object> changes,
        string? ipAddress)
    {
        return new AuditLog
        {
            EntityType = entityType,
            EntityId = entityId,
            Action = action,
            UserId = userId,
            UserName = userName,
            Timestamp = DateTime.UtcNow,
            Changes = changes,
            IpAddress = ipAddress
        };
    }

    /// <summary>
    /// ファクトリメソッド：UPDATE 操作用
    /// </summary>
    public static AuditLog CreateForUpdate(
        string entityType,
        string entityId,
        string userId,
        string userName,
        Dictionary<string, object> oldValues,
        Dictionary<string, object> newValues,
        string? ipAddress)
    {
        return new AuditLog
        {
            EntityType = entityType,
            EntityId = entityId,
            Action = AuditAction.UPDATE,
            UserId = userId,
            UserName = userName,
            Timestamp = DateTime.UtcNow,
            OldValues = oldValues,
            NewValues = newValues,
            IpAddress = ipAddress
        };
    }

    /// <summary>
    /// ファクトリメソッド：DELETE 操作用
    /// </summary>
    public static AuditLog CreateForDelete(
        string entityType,
        string entityId,
        string userId,
        string userName,
        Dictionary<string, object> oldValues,
        string? reason,
        string? ipAddress)
    {
        return new AuditLog
        {
            EntityType = entityType,
            EntityId = entityId,
            Action = AuditAction.DELETE,
            UserId = userId,
            UserName = userName,
            Timestamp = DateTime.UtcNow,
            OldValues = oldValues,
            Reason = reason,
            IpAddress = ipAddress
        };
    }

    /// <summary>
    /// サマリー文字列を生成
    /// </summary>
    public string GetSummary()
    {
        return $"{EntityType} {EntityId} を{Action.GetDisplayName()}";
    }
}

File: app/fsharp/AccountingSystem.Domain/Models/AuditLog.fs
namespace AccountingSystem.Domain.Models

open System
open AccountingSystem.Domain.Types

/// <summary>
/// 監査ログドメインモデル（不変）
/// 財務会計システムにおける法的要件（5W1H）を満たすため、
/// すべての重要な操作を記録します。
/// </summary>
type AuditLog = {
    /// 監査ログID（主キー）
    Id: int64 option
    /// エンティティ種別（例: "Account", "Journal"）
    EntityType: string
    /// エンティティID（例: "1001", "J-2024-0001"）
    EntityId: string
    /// 操作種別
    Action: AuditAction
    /// 操作ユーザーID
    UserId: string
    /// 操作ユーザー名
    UserName: string
    /// 発生日時
    Timestamp: DateTime
    /// 変更前の値（JSON形式）
    OldValues: string option
    /// 変更後の値（JSON形式）
    NewValues: string option
    /// 変更内容（JSON形式）
    Changes: string option
    /// 変更理由
    Reason: string option
    /// IPアドレス
    IpAddress: string option
    /// ユーザーエージェント
    UserAgent: string option
}

module AuditLog =
    /// <summary>
    /// CREATE操作用のファクトリ関数
    /// </summary>
    let create
        (entityType: string)
        (entityId: string)
        (userId: string)
        (userName: string)
        (changes: string)
        (ipAddress: string option) : AuditLog =
        {
            Id = None
            EntityType = entityType
            EntityId = entityId
            Action = AuditAction.Create
            UserId = userId
            UserName = userName
            Timestamp = DateTime.UtcNow
            OldValues = None
            NewValues = Some changes
            Changes = Some changes
            Reason = None
            IpAddress = ipAddress
            UserAgent = None
        }

    /// <summary>
    /// UPDATE操作用のファクトリ関数
    /// </summary>
    let createForUpdate
        (entityType: string)
        (entityId: string)
        (userId: string)
        (userName: string)
        (oldValues: string)
        (newValues: string)
        (ipAddress: string option) : AuditLog =
        {
            Id = None
            EntityType = entityType
            EntityId = entityId
            Action = AuditAction.Update
            UserId = userId
            UserName = userName
            Timestamp = DateTime.UtcNow
            OldValues = Some oldValues
            NewValues = Some newValues
            Changes = None
            Reason = None
            IpAddress = ipAddress
            UserAgent = None
        }

    /// <summary>
    /// DELETE操作用のファクトリ関数
    /// </summary>
    let createForDelete
        (entityType: string)
        (entityId: string)
        (userId: string)
        (userName: string)
        (oldValues: string)
        (reason: string option)
        (ipAddress: string option) : AuditLog =
        {
            Id = None
            EntityType = entityType
            EntityId = entityId
            Action = AuditAction.Delete
            UserId = userId
            UserName = userName
            Timestamp = DateTime.UtcNow
            OldValues = Some oldValues
            NewValues = None
            Changes = None
            Reason = reason
            IpAddress = ipAddress
            UserAgent = None
        }

    /// <summary>
    /// サマリー文字列を生成
    /// </summary>
    let getSummary (auditLog: AuditLog) : string =
        $"{auditLog.EntityType} {auditLog.EntityId} を{AuditAction.getDisplayName auditLog.Action}"
